stdlib/public/runtime/ImageInspectionCOFF.cpp

```diff
@@ -99,27 +184,13 @@ void swift::initializeTypeFieldLookup() {
 SWIFT_RUNTIME_EXPORT
 void swift_addNewDSOImage(const void *addr) {
   const swift::MetadataSections *sections =
-      static_cast<const swift::MetadataSections *>(addr);
+  static_cast<const swift::MetadataSections *>(addr);
 
   record(sections);
 
-  const auto &protocols_section = sections->swift5_protocols;
-  const void *protocols =
-      reinterpret_cast<void *>(protocols_section.start);
-  if (protocols_section.length)
-    addImageProtocolsBlockCallback(protocols, protocols_section.length);
-
-  const auto &protocol_conformances = sections->swift5_protocol_conformances;
-  const void *conformances =
-      reinterpret_cast<void *>(protocol_conformances.start);
-  if (protocol_conformances.length)
-    addImageProtocolConformanceBlockCallback(conformances,
-                                             protocol_conformances.length);
-
-  const auto &type_metadata = sections->swift5_type_metadata;
-  const void *metadata = reinterpret_cast<void *>(type_metadata.start);
-  if (type_metadata.length)
-    addImageTypeMetadataRecordBlockCallback(metadata, type_metadata.length);
+  registerImageProtocolsSection(sections->swift5_protocols);
+  registerImageProtocolConformanceSection(sections->swift5_protocol_conformances);
+  registerImageTypeMetadataRecordSection(sections->swift5_type_metadata);
```

Comment:
I really would prefer that we keep this consistent across ELF and COFF.  Why can't we do the adjustment inside the scanning itself?  That is, sink this into the callbacks.

Reply:
The scanning is only necessary on Windows to work around the behavior of the MSVC linker in incremental linking mode. The callbacks on the other hand are defined in the platform independent part (e.g. MetadataLookup.cpp).

So I don't think that we should pull Windows specific code into the parts of the runtime that affect Windows, Linux and macOS. This just increases the required testing and dependency surface while keeping the code here In the COFF specific implementation files means that only Windows specific code is impacted and requires testing.

Reply:
I've looked through the runtime code to see which files exactly would be impacted if we would move the scanning into the callbacks. Impacted files:

MetadataLookup.cpp
ProtocolConformance.cpp

which means that we would have to define the scanning code in some .cpp file and then declare its prototype in some .h so that the callback-defining files could use the scanning code.

It would be a more complex solution and it would split up the code that deals with the incremental linking problem over multiple files while the way I did it in my patch keeps everything together and in one place making it easier for other people in the future to understand what's going on and why we do it.